code/modules/power/apc/apc.dm

```diff
@@ -0,0 +1,1063 @@
+	var/datum/wires/apc/wires = null
+	/// Can the APC recieve/transmit power? Determined by the condition of the 2 Main Power Wires
+	var/shorted = FALSE
+	/// Is the APC on and transmitting power to enabled breakers? Think of this var as the master breaker for the APC
+	var/operating = TRUE
+	/// The current charging mode of the APC: not charging, charging, fully charged
+	var/charging = APC_NOT_CHARGING
+	/// A counter var used to help determine if the APC has not been charging long enough to justify not performing certain auto setting such as turning all channels back on
+	var/longtermpower = 10
+	/// State of the APC Cover - Closed, Open, or Off
+	var/opened = APC_CLOSED
+	/// Can silicons access this APC?
+	var/aidisabled = FALSE
+	var/electronics_state = APC_ELECTRONICS_NONE
+
+	/*** APC Settings Vars ***/
+	/// The current setting for the lighting channel
+	var/lighting_channel = CHANNEL_SETTING_AUTO_ON
+	/// The current setting for the equipment channel
+	var/equipment_channel = CHANNEL_SETTING_AUTO_ON
+	/// The current setting for the environment channel
+	var/environment_channel = CHANNEL_SETTING_AUTO_ON
+	/// Is the APC cover locked? i.e cannot be opened?
+	var/coverlocked = TRUE
+	/// Is the APC User Interface locked (prevents interaction)? Will not prevent silicons or admin observers from interacting
+	var/locked = TRUE
+	/// If TRUE, the APC will automatically draw power from connect terminal, if FALSE it will not charge
+	var/chargemode = TRUE
+	/// Counter var, ticks up when the APC recieves power from terminal and resets to 0 when not charging, used for the `var/charging` var
+	var/chargecount = 0
+	var/report_power_alarm = TRUE
+
+	var/shock_proof = FALSE //if set to FALSE, this APC will not arc bolts of electricity if it's overloaded.
+
+	// Nightshift
+	var/nightshift_lights = FALSE
+	var/last_nightshift_switch = 0
+
+	/// Used to determine if emergency lights should be on or off
+	var/emergency_power = TRUE
+	var/emergency_power_timer
+	var/emergency_lights = FALSE
+
+	/// settings variable for having the APC auto use certain power channel settings
+	var/autoflag = APC_AUTOFLAG_ALL_OFF		// 0 = off, 1= eqp and lights off, 2 = eqp off, 3 = all on.
+
+	/*** APC Malf AI Vars ****/
+	var/malfhack = FALSE //New var for my changes to AI malf. --NeoFite
+	var/mob/living/silicon/ai/malfai = null //See above --NeoFite
+	var/mob/living/silicon/ai/occupier = null
+	/// Was this APC built instead of already existing? Used for malfhack to keep borgs from building apcs in space
+	var/constructed = FALSE
+	var/overload = 1 //used for the Blackout malf module
+
+	/*** APC Overlay Vars ***/
+	var/update_state = -1
+	var/update_overlay = -1
+	var/global/status_overlays = FALSE
+	var/updating_icon = FALSE
+	var/global/list/status_overlays_lock
+	var/global/list/status_overlays_charging
+	var/global/list/status_overlays_equipment
+	var/global/list/status_overlays_lighting
+	var/global/list/status_overlays_environ
+	var/keep_preset_name = FALSE
```

Comment:
What the hell is this disgusting crap? APC code is wild.

Reply:
yeah... it's pretty useful but its so hard to figure out. It's all just performance-saving stuff for overlays.